In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [3]:
df = pd.read_csv("air_fryers_clean_brand_year.csv")
df.head()

,category,year,brand,purchase_count,product_count,avg_price,avg_rating,compact_share,dual_basket_share,oven_style_share,rotisserie_share,window_share,market_purchases,brand_share,log_brand_share
0,air_fryers,2019,chefman,1146,10,72.963695,4.434119,1.000000,0.0,0.780977,0.243455,0.184119,15076,0.076015,-2.576826
1,air_fryers,2019,cosori,11,2,159.990000,4.581818,1.000000,0.0,0.090909,0.090909,0.000000,15076,0.000730,-7.222964
2,air_fryers,2019,cuisinart,1616,22,229.465274,4.481312,0.993812,0.0,0.889851,0.000000,0.000000,15076,0.107190,-2.233150
3,air_fryers,2019,dash,3011,19,55.176333,4.390767,1.000000,0.0,0.973431,0.000000,0.000000,15076,0.199721,-1.610832
4,air_fryers,2019,gowise usa,4405,45,83.575551,4.552259,0.999773,0.0,0.129398,0.128490,0.000000,15076,0.292186,-1.230364


In [6]:
y = df["log_brand_share"]

brand_dummies = pd.get_dummies(df["brand"],
                prefix="brand", drop_first=True, dtype=int)
year_dummies = pd.get_dummies(df["year"].astype(str),
                prefix="year", drop_first=True, dtype=int)

feature_cols = [col for col in df.columns if col not in 
                ["category", "brand", "year", "log_brand_share", 
                 "brand_share", "avg_price", "avg_rating",
                 "purchase_count", "product_count", "market_purchases"]]

X = pd.concat(
    [df[["avg_price", "avg_rating"] + feature_cols],
     brand_dummies,
     year_dummies],
    axis=1,
)

model = LinearRegression()
model.fit(X, y)

predicted_log_share = model.predict(X)
r2 = r2_score(y, predicted_log_share)

coef_table = pd.DataFrame({
    "feature": X.columns,
    "coefficient": model.coef_
})

print("R-squared:", r2)
coef_table

R-squared: 0.7634539500914359


,feature,coefficient
0,avg_price,-0.037668
1,avg_rating,0.287517
2,compact_share,9.815304
3,dual_basket_share,-9.509686
4,oven_style_share,1.941774
5,rotisserie_share,-5.674054
6,window_share,12.880298
7,brand_cosori,2.551946
8,brand_cuisinart,6.422436
9,brand_dash,0.176655


## 2. Answers to Part 2 - Demand Estimation 

1. **What is the estimated price coefficient?**  
   The estimated price coefficient is -0.037668.

2. **Is it negative? Why is that important?**  
   Yes, the estimated price coefficient is negative. This is important because proves that the law of demand is true. As the price for a brand increases, the demand for brand share decreases. 

3. **Which product features are associated with higher demand?**  
   Product features compact_share, window_share, and oven_style_share are all associated with higher demand due to their positive coefficients of 9.82, 12.88, and 1.94 respectively. However, window_share has the strongest positive effect due to the largest, positive coefficient value.

4. **Which brand dummy coefficients are largest?**  
   brand_cuisinart and brand_ninja's brand dummy coefficients were the largest with values of 6.42 and  5.84, meaning these brands have significantly higher demand relative to the brand, chefman, which was dropped to serve as the baseline reference for our model.

5. **Which year dummy coefficients are largest?**  
   year_2020 had the largest dummy coefficients. of 0.119 was the largest, meaning demand was highest in 2020 relative to the reference year of 2019. 

6. **What is the model's R²?**  
   The model's R² = 0.763, which means that the model explains approximately 76% of the variation in log brand share.